### PR TITLE
Adds TemplateData and TemplateRequests

### PR DESCRIPTION
**Pull request by Redstoneguy129**
**Justification:** 
discord/discord-api-docs#1712
discord/discord-api-docs#2144

**Issue**
https://github.com/Discord4J/Discord4J/issues/791

I will be making a PR to Discord4J main repo after this with the Templates.

### CODE REVIEW

**Comment by quanticc** [comment]
Is this an API v8 feature only?

If v6 is supported, this PR can be submitted to 1.5.x branch instead

**Comment by Redstoneguy129** [comment]
I'm not used to the API but seeing that it's a new feature on discord I think it's safe to say its API v8.
https://discord.com/developers/docs/resources/template

**Comment by Redstoneguy129** [comment]
https://discord.com/api/v6/ and https://discord.com/api/v8/ are both supported

File: src/main/java/discord4j/discordjson/json/RoleData.java [hunk 1]
```diff
@@ -18,17 +20,17 @@
 
     String name();
 
-    int color();
+    Integer color();
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/RoleData.java:23]: Why did you change this? If there is no reason, primitives (boolean, int, long...) are preferable over objects

Comment (Redstoneguy129) [reply src/main/java/discord4j/discordjson/json/RoleData.java:23]: My IDE didn't like it for some weird reason. I forgot to change it back when making this PR sorry.

File: src/main/java/discord4j/discordjson/json/SerializedSourceGuildData.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    String name();
+    Optional<String> description();
+    String region();
+    @JsonProperty("verification_level")
+    Integer verificationLevel();
+    @JsonProperty("default_message_notifications")
+    Integer defaultMessageNotifications();
+    @JsonProperty("explicit_content_filter")
+    Integer explicitContentFilter();
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/SerializedSourceGuildData.java:28]: Can you use primitives here too please?

**Review by Shadorc — CHANGES_REQUESTED** [review]

File: src/main/java/discord4j/discordjson/json/SerializedSourceGuildData.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @JsonProperty("preferred_locale")
+    String preferredLocale();
+    @JsonProperty("afk_timeout")
+    int afkTimeout();
+    List<RoleData> roles();
+    List<ChannelData> channels();
+    @JsonProperty("afk_channel_id")
+    Optional<Integer> afkChannelId();
+    @JsonProperty("system_channel_id")
+    int systemChannelId();
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/SerializedSourceGuildData.java:38]: IDs are deserialized either as string or long, int will overflow. Discord4j chose string (and later long but this change is not done yet so I think String for now is the corrrect way)

**Review by Shadorc — CHANGES_REQUESTED** [review]

**Review by quanticc — CHANGES_REQUESTED** [review]
Please target 1.5.x branch so we can add this feature to Discord4J 3.1.x

**Comment by quanticc** [comment]
I recommend in the future that all PRs start as a branch in your own repo, that way you can easily change target branch by:

```
git rebase --onto 1.5.x master your-feature-branch
```

means: bring all commits from `master` to `your-feature-branch` on top of `1.5.x`

File: src/main/java/discord4j/discordjson/json/RoleData.java [hunk 4]
```diff
@@ -22,11 +24,11 @@
 
     boolean hoist();
 
-    int position();
+    Possible<Optional<Integer>> position();
```
Comment (quanticc) [inline src/main/java/discord4j/discordjson/json/RoleData.java:27]: I need to review these field type changes (position and managed). https://discord.com/developers/docs/topics/permissions#role-object don't indicate this is a nullable value, optional field.

Maybe a different RoleData class is a better fit. if RoleData is our full Role object, but this new RoleTemplateData partially covers it. My suggestion is to:

- Create a new RoleDataFields interface with fields common to both (check existing interfaces ending in -Fields)
- Make RoleData extend RoleDataFields, and adding the position + managed fields
- Create a new RoleTemplateData that extends RoleDataFields, only adding the static builder method standard to our data objects

Let me know what you think about it and if it's right or not. Also @darichey @danthonywalker please take a look

**Review by quanticc — CHANGES_REQUESTED** [review]
Added a change request to RoleData, to avoid a field type change.

About SerializedSourceGuildData, it's fine but after I merge #42, I'm going to consider factoring out common fields to other Guild*Fields interfaces

File: src/main/java/discord4j/discordjson/json/RoleData.java (hunk 4 above)
Comment (Redstoneguy129) [reply src/main/java/discord4j/discordjson/json/RoleData.java:27]: I'll create a RoleTemplateData that extends RoleData. When using a template the position object didn't exist so i thought Discord changed how RoleData works.

Comment (quanticc) [reply src/main/java/discord4j/discordjson/json/RoleData.java:27]: RoleData will have the same fields and you want to exclude `position` and `managed` from RoleTemplateData, so make sure the latter extends an interface with only the common fields.

Comment (Redstoneguy129) [reply src/main/java/discord4j/discordjson/json/RoleData.java:27]: Yes i just realised that
